packages/vite/src/node/optimizer/scan.ts

```diff
@@ -642,36 +1018,27 @@ function esbuildScanPlugin(
           contents = config.esbuild.jsxInject + `\n` + contents
         }
 
-        const loader =
-          config.optimizeDeps?.esbuildOptions?.loader?.[`.${ext}`] ||
-          (ext as Loader)
+        const loader = ext as Loader
+
+        if (loader !== 'js') {
+          contents = (await transform(contents, { loader })).code
+        }
```

Comment:
I was checking a test failure in `playground/tsconfig-json` and the error happens during rolldown scan due to esbuild transform not having `experimentalDecorators` enabled.

https://github.com/rolldown/vite/blob/9fdf94a3e51a9a29db3bfac0a50d943478361821/playground/tsconfig-json/__tests__/tsconfig-json.spec.ts#L24-L32

I think this issue didn't surface previously because the transform was tucked inside `if (contents.includes('import.meta.glob'))` and `doTransformGlobImport`. Maybe would it be okay to keep that structure?

Reply:
Here cuased by the `transform` not pass the `tsconfig.json` to esbuild, I lost it and fix it at now.

Reply:
I think `loadTsconfigJsonForFile` part used in `prepareEsbuildScanner` was a different. Technically Vite still has a bug since this `transform` is independently called from scanner esbuild context, but it just happens that the issue doesn't manifest since `transform` is only called when the code includes `import.meta.glob`.

Nonetheless, your fix https://github.com/rolldown/vite/pull/24/commits/0f009c945aff43f6560d1a9cef11eb5d97edf83c makes more sense. Thanks :+1: I'm closing my PR https://github.com/rolldown/vite/pull/32